src/Model.php

```diff
@@ -871,6 +879,118 @@ public function offsetUnset($name)
 
     // }}}
 
+    // {{{ UserAction support
+
+    /**
+     * Register new user action for this model. By default UI will allow users to trigger actions
+     * from UI.
+     *
+     * @param $name
+     * @param null  $callback
+     * @param array $arguments
+     * @param array $defaults
+     *
+     * @throws \atk4\core\Exception
+     *
+     * @return UserAction\Action
+     */
+    public function addAction($name, $defaults = [])
+    {
+        if (is_callable($defaults)) {
+            $defaults = ['callback'=>$defaults];
+        }
+
+        if (!isset($defaults['title'])) {
+            $s = $name;
+
+            $s = preg_split('/[\\\\_]/', $s, -1, PREG_SPLIT_NO_EMPTY);
+            $s = array_map('trim', $s);
+            $s = ucwords(implode(' ', $s));
+
+            $defaults['caption'] = $s;
+        }
+
+        /** @var UserAction\Action $action */
+        $action = $this->factory($this->_default_action, $defaults);
+
+        $this->add($action, 'action:'.$name);
+
+        return $action;
+    }
+
+    /**
+     * Returns list of actions for this model.
+     *
+     * @return array
+     */
+    public function getActions($scope = null)
+    {
+        $actions = array_filter(array_keys($this->elements), function ($var) use ($scope) {
+            if (stripos($var, 'action:') !== 0) {
+                return false;
+            }
+
+            if ($scope !== null && $this->getElement($var)->scope !== $scope) {
+                return false;
+            }
+
+            return true;
+        });
+        $res = [];
+
+        foreach ($actions as $action) {
+            $a = $this->getElement($action);
+
+            if ($a->system) {
+                continue;
+            }
+
+            $res[str_replace('action:', '', $action)] = $a;
+        }
+
+        return $res;
+    }
+
+    /**
+     * Finds a user action with a corresponding name. Returns false if action not found. Similar
+     * to hasElement() but with extra checks to make sure it's certainly a action you are
+     * getting.
+     *
+     * @param string $name
+     *
+     * @return UserAction\Action|false
+     */
+    public function hasAction($name)
+    {
+        $actions = array_filter(array_keys($this->elements), function ($var) use ($name) {
+            return stripos($var, 'action:'.$name) === 0;
```

Comment:
it would be good to refactor this and step out of hasAction method as soon as first matching action is found.

Reply:
yeh, i was contemplating about having alternative to "elements" for all those.. Not sure how though.